### ISSUE TITLE
Can the new version of (not yet checksummed) obtained by `spack versions xxx` automatically adapt to the correct source code pull path during installation?

### ISSUE THREAD

**Issue by simon28li** [LLNL/spack#23810]
Can the new version of (not yet checksummed) obtained by `spack versions xxx` automatically adapt to the correct source code pull path during installation?
### Rationale

The problems I encountered are similar to this issue[#14129](https://github.com/spack/spack/issues/14129).
### Description

```
[root@localhost ~]# spack checksum gperftools
==> Found 11 versions of gperftools:

  2.9.1   https://github.com/gperftools/gperftools/releases/download/gperftools-2.9.1/gperftools-2.9.1.tar.gz
  2.9.0   https://github.com/gperftools/gperftools/releases/download/gperftools-2.9/gperftools-2.9.0.tar.gz
  2.9     https://github.com/gperftools/gperftools/archive/refs/tags/gperftools-2.9.tar.gz
  2.8.90  https://github.com/gperftools/gperftools/releases/download/gperftools-2.8.90/gperftools-2.8.90.tar.gz
  2.8.1   https://github.com/gperftools/gperftools/releases/download/gperftools-2.8.1/gperftools-2.8.1.tar.gz
  2.8     https://github.com/gperftools/gperftools/releases/download/gperftools-2.8/gperftools-2.8.tar.gz
  2.7.90  https://github.com/gperftools/gperftools/releases/download/gperftools-2.7.90/gperftools-2.7.90.tar.gz
  2.7     https://github.com/gperftools/gperftools/releases/download/gperftools-2.7/gperftools-2.7.tar.gz
  2.6.90  https://github.com/gperftools/gperftools/releases/download/gperftools-2.6.90/gperftools-2.6.90.tar.gz
  ...
  2.6.2   https://github.com/gperftools/gperftools/releases/download/gperftools-2.6.2/gperftools-2.6.2.tar.gz
```
My goal is to get which new versions are available, and then test the compilation and installation of each version, but some versions report an error when downloading the source code as follows.
```
[root@localhost ~]# spack install -v gperftools@2.9
==> Warning: Missing a source id for gperftools@2.9
[+] /home/spack/opt/spack/linux-centos8-aarch64/gcc-8.3.1/libunwind-1.5.0-e37egh35bf6ms4dkoulga7mjvbfvc474
==> Installing gperftools-2.9-d4agpypmjylcj2kmlxaqbf3xeuovtefc
==> No binary for gperftools-2.9-d4agpypmjylcj2kmlxaqbf3xeuovtefc found: installing from source
==> Warning: Missing a source id for gperftools@2.9
==> Warning: There is no checksum on file to fetch gperftools@2.9 safely.
==>   Fetch anyway? [y/N] y
==> Error: FetchError: All fetchers failed

/home/spack/lib/spack/spack/package.py:1360, in do_fetch:
       1357
       1358        self.stage.create()
       1359        err_msg = None if not self.manual_download else self.download_instr
  >>   1360        start_time = time.time()
       1361        self.stage.fetch(mirror_only, err_msg=err_msg)
       1362        self._fetch_time = time.time() - start_time
       1363

```
I have observed that some versions obtained by the `spack versions` command do not match the corresponding source code path regularly, resulting in not downloading the source code during installation. Is there a unified way to automatically adapt to the correct source code download path?

```
* **Spack:** 0.16.1
* **Python:** 3.6.8
* **Platform:** linux-centos8-aarch64
* **Concretizer:** original
```

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by simon28li** [comment]
@adamjstewart Excuse me. I'm very sorry. Could you give me some conclusions? Our own project has some requirements related to this and has encountered bottlenecks.

**Comment by adamjstewart** [comment]
This is a longstanding issue in Spack. To be fair, Spack is one of the few package managers that attempts to handle this kind of situation. In most package managers, the package contains only a single version, and the URL for that version. In Spack, not only can you list multiple versions, but you can also list multiple URLs, or list a single URL and let Spack guess.

In the case of GitHub packages like this, the problem is that there are multiple tarballs for the same version, and it isn't clear which one is "better". I think most people would agree that consistency is better, and what we have right now is certainly not consistent, as it's relatively random whether you get the GitHub tagged source code or a re-uploaded tarball.

In the case of gperftools in particular, I would argue that this is a nightmare they brought on themselves by lacking consistency. For example, they mix X.Y versions like 2.9 with X.Y.Z versions like 2.9.1. Then, depending on which X.Y release you're looking at, they change the tarball names:
https://github.com/gperftools/gperftools/releases/download/gperftools-2.9/gperftools-2.9.0.tar.gz
https://github.com/gperftools/gperftools/releases/download/gperftools-2.8/gperftools-2.8.tar.gz

Basically, what this boils down to for gperftools in particular, is that it really isn't possible to guess what the URL might be, and there are 4+ to choose from for every version, so there's not much we can do. For other GitHub packages, I definitely think we can do better, and there are already dozens of open issues on this, but I haven't yet figured out a better solution than what we currently have.

Hope this helps!

**Comment by simon28li** [comment]
@adamjstewart Thank you very much for your reply.
